### ISSUE TITLE
#64 ⁃ Ensure direct chats don't have more than two accounts before sending E2EE verification messages

### ISSUE THREAD

**Issue by famedly-bot** [famedly/matrix-dart-sdk#248]
In GitLab by @Mikaela on Nov 13, 2021, 11:42

Soru requested me to report that verification messages should ensure direct chat doesn't have more than two users as it just kept failing due to sending messages to room of more users with other devices sending reject messages. I commonly have three or more accounts in direct chats and consider private rooms between just a few people as direct chats.

**Comment by Mar-Ga** [comment]
@famedly/instant-messaging is this still an issue or can we close this

**Comment by nico-famedly** [comment]
No, this is still an issue: https://github.com/famedly/matrix-dart-sdk/blob/31a52cb463b404b7a7ade3e3657f09770185a573/lib/src/utils/device_keys_list.dart#L80

That call to startDirectChat might return an existing DM where someone else got invited to.

**Comment by krille-chan** [comment]
So to fix this I would suggest that we change the method `startDirectChat()` with a new optional parameter like this:

```dart
  /// Returns an existing direct room ID with this user or creates a new one.
  /// By default encryption will be enabled if the client supports encryption
  /// and the other user has uploaded any encryption keys.
  Future<String> startDirectChat(
    String mxid, {
    bool? enableEncryption,
    List<StateEvent>? initialState,
    bool waitForSync = true,
    Map<String, dynamic>? powerLevelContentOverride,
    CreateRoomPreset? preset = CreateRoomPreset.trustedPrivateChat,

    /// Does not return existing DM rooms which have more than two members.
    bool ensureIsOnlyTwoMembers = false,
  })
```

and then set `ensureIsOnlyTwoMembers: true` when using this method in `startVerification()` method.
